sql/analyzer/indexed_joins.go

```diff
@@ -0,0 +1,633 @@
+		newRight, err := replaceTableAccessWithIndexedAccess(node.Right, append(schema, node.Left.Schema()...), scope, joinIndexes, exprAliases, tableAliases)
+		if err != nil {
+			return nil, err
+		}
+		return node.WithChildren(node.Left, newRight)
+	default:
+		// For an unhandled node type, just skip this transformation
+		return node, nil
+	}
+}
+
+func replanJoin(node plan.JoinNode, a *Analyzer, joinIndexes joinIndexesByTable, ) (sql.Node, error) {
+	// Inspect the node for eligibility. The join planner rewrites the tree beneath this node, and for this to be correct
+	// only certain nodes can be below it.
+	eligible := true
+	plan.Inspect(node, func(node sql.Node) bool {
+		switch node.(type) {
+		case plan.JoinNode, *plan.ResolvedTable, *plan.TableAlias, nil:
+		default:
+			a.Log("Skipping join replanning because of incompatible node: %T", node)
+			eligible = false
+		}
+		return true
+	})
+
+	if !eligible {
+		return node, nil
+	}
+
+	// Collect all tables and find an access order for them
+	tables := lexicalTableOrder(node)
+	tablesByName := byLowerCaseName(tables)
+	tableOrder := orderTables(tables, tablesByName, joinIndexes)
+
+	// Then use that order to construct a join tree
+	joinTree := buildJoinTree(tableOrder, joinIndexes.flattenJoinConds())
+
+	// This shouldn't happen, but better to fail gracefully if it does
+	if joinTree == nil {
+		return node, nil
+	}
+
+	joinNode := joinTreeToNodes(joinTree, tablesByName)
+
+	return joinNode, nil
+}
+
+// lexicalTableOrder returns the names of the tables under the join node given, in the original lexical order. This is
+// possible to reconstruct because the parser assembles joins in left-bound fashion, so that the first two tables are
+// the most deeply nested node on the left branch. Each additional join wraps this original join node, with the
+// original on the left and the new table being joined on the right.
+func lexicalTableOrder(node sql.Node) []NameableNode {
+	switch node := node.(type) {
+	case *plan.TableAlias:
+		return []NameableNode{node}
+	case *plan.ResolvedTable:
+		return []NameableNode{node}
+	case plan.JoinNode:
+		return append(lexicalTableOrder(node.LeftBranch()), lexicalTableOrder(node.RightBranch())...)
+	default:
+		panic(fmt.Sprintf("unexpected node type: %t", node))
+	}
+}
+
+func joinTreeToNodes(tree *joinSearchNode, tablesByName map[string]NameableNode) sql.Node {
+	if len(tree.table) > 0 {
+		return tablesByName[tree.table]
+	}
+
+	left := joinTreeToNodes(tree.left, tablesByName)
+	right := joinTreeToNodes(tree.right, tablesByName)
+	return plan.NewIndexedJoin(left, right, tree.joinCond.joinType, tree.joinCond.cond)
+}
+
+// createPrimaryTableExpr returns a slice of expressions to be used when evaluating a row in the primary table to
+// assemble a lookup key in a secondary table. Column expressions must match the declared column order of the index.
+func createIndexLookupKeyExpression(
```

Comment:
Starting from here, comments are either missing the function name or have the wrong name.

Reply:
Fixed.